### PR TITLE
Improve deployment address checking

### PR DESCRIPTION
**Pull request by nventuro**
# Description

This is a very simple way to add checks for contracts that are deployed by other contracts, either via `create` or `create2`, and therefore not possible to check via our standard mechanism (which expects contracts to be deployed by EOAs).

Since these contracts' addresses are stored via `save()`, we can introduce a check there to compare what is being saved to what is stored as the task's output.

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

N/A

### CODE REVIEW

File: pkg/deployments/src/task.ts [hunk 1]
```diff
@@ -286,14 +286,39 @@ export default class Task {
   }
 
   save(output: RawOutput): void {
+    const parsedOutput = this._parseRawOutput(output);
+
+    if (this.mode === TaskMode.CHECK) {
+      // `save` is only called when by `deploy` (which only happens in LIVE mode), or manually for contracts that are
```
Comment (EndymionJkb) [inline pkg/deployments/src/task.ts:292]:
```suggestion
      // `save` is called either by `deploy` (which only happens in LIVE mode), or manually for contracts that are
```

File: pkg/deployments/src/task.ts [hunk 2]
```diff
@@ -286,14 +286,39 @@ export default class Task {
...
+      // deployed by other contracts (e.g. Batch Relayer Entrypoints). Therefore, by testing for CHECK mode we can
+      // indentify this second type of contracts, and check them by comparing the saved address to the address that the
+      // task would attempt to save.
```
Comment (EndymionJkb) [inline pkg/deployments/src/task.ts:295]:
```suggestion
      // identify this second type of contracts, and check them by comparing the saved address to the address that the
      // task is attempting to save.
```

File: pkg/deployments/src/task.ts [hunk 3]
```diff
@@ -286,14 +286,39 @@ export default class Task {
   }
 
   save(output: RawOutput): void {
```
Comment (TomAFrench) [inline pkg/deployments/src/task.ts:288]: I'd break this up into two separate private functions which we select between using the taskmode.

```
 save(output: RawOutput): void {
   if (this.mode === TaskMode.LIVE) {
     this._save(output);
   } else if (this.mode === TaskMode.CHECK) {
     this._smortCheck(output);
   }
```

where `_save()` is the previous function and `_smortCheck()` is the new code.

File: pkg/deployments/src/task.ts (hunk 2 above)
Comment (TomAFrench) [inline pkg/deployments/src/task.ts:295]:
```suggestion
      // `save` is only called by `deploy` (which only happens in LIVE mode), or manually for contracts that are
      // deployed by other contracts (e.g. Batch Relayer Entrypoints). Therefore by testing for CHECK mode we can
      // identify this second type of contracts, and check them by comparing the saved address to the address that the
      // task would attempt to save.
```

This should also get moved up into the dispatcher function described above.

**Review by jubeira — COMMENTED** [review]
Cool!
I agree with Tom's proposal about splitting into private functions :+1:

File: pkg/deployments/src/task.ts [hunk 4]
```diff
@@ -286,14 +286,39 @@ export default class Task {
...
+    const parsedOutput = this._parseRawOutput(output);
+
+    if (this.mode === TaskMode.CHECK) {
+      // `save` is only called when by `deploy` (which only happens in LIVE mode), or manually for contracts that are
+      // deployed by other contracts (e.g. Batch Relayer Entrypoints). Therefore, by testing for CHECK mode we can
+      // indentify this second type of contracts, and check them by comparing the saved address to the address that the
+      // task would attempt to save.
+
+      for (let name of Object.keys(parsedOutput)) {
```
Comment (TomAFrench) [inline pkg/deployments/src/task.ts:297]:
```suggestion
      for (const name of Object.keys(parsedOutput)) {
```

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**